services/medco/libmedco/survey_store.go

```diff
@@ -0,0 +1,199 @@
+package libmedco
+
+import (
+	"fmt"
+)
+
+// DefaultGroup defines the default grouping key and is used when a survey consists of an aggregation only (no grouping).
+const DefaultGroup = GroupingKey("DefaultGroup")
+
+// SurveyResult represents a result from a survey which is made of two ciphervector encrypted for the querier.
+type SurveyResult struct {
+	GroupingAttributes    CipherVector
+	AggregatingAttributes CipherVector
+}
+
+// SurveyStore contains all the elements of a survey, it consists of the data structure that each cothority has to
+// maintain locally to perform a collective survey.
+type SurveyStore struct {
+	ClientResponses    []ClientResponse
+	DeliverableResults []SurveyResult
+
+	// AggregatingAttributes is a map containing all the aggregating attributes of the server clients and a
+	// corresponding unique ID.
+	AggregatingAttributes map[TempID]CipherVector
+
+	// LocGroupingAggregating contains the results of the local aggregation.
+	LocGroupingAggregating map[GroupingKey]CipherVector
+
+	// AfterAggrProto & LocGroupingGroups are the results of the split of LocGroupingAggregating, first one contains
+	// aggregating attributes and the second the grouping attributes.
+	AfterAggrProto    map[GroupingKey]CipherVector
+	LocGroupingGroups map[GroupingKey]GroupingAttributes
+
+	// GroupedDeterministicGroupingAttributes & GroupedAggregatingAttributes contain results of the grouping
+	// before they are key switched and combined in the last step (key switching).
+	GroupedDeterministicGroupingAttributes map[TempID]GroupingAttributes
+	GroupedAggregatingAttributes           map[TempID]CipherVector
+
+	lastID uint64
+}
+
+// NewSurveyStore is the store constructor.
+func NewSurveyStore() *SurveyStore {
+	return &SurveyStore{
+		AggregatingAttributes: make(map[TempID]CipherVector),
+
+		LocGroupingAggregating: make(map[GroupingKey]CipherVector),
+		LocGroupingGroups:      make(map[GroupingKey]GroupingAttributes),
+		AfterAggrProto:         make(map[GroupingKey]CipherVector),
+
+		GroupedDeterministicGroupingAttributes: make(map[TempID]GroupingAttributes),
+		GroupedAggregatingAttributes:           make(map[TempID]CipherVector),
+	}
+}
+
+// InsertClientResponse handles the local storage of a new client response in aggregation or grouping cases.
+func (s *SurveyStore) InsertClientResponse(cr ClientResponse) {
+	if len(cr.ProbabilisticGroupingAttributes) == 0 { //only aggregation, no grouping
+		mapValue, ok := s.LocGroupingAggregating[DefaultGroup]
+		if !ok {
+			mapValue = cr.AggregatingAttributes
+		} else {
+			mapValue = *NewCipherVector(len(mapValue)).Add(mapValue, cr.AggregatingAttributes)
+		}
+		s.LocGroupingAggregating[DefaultGroup] = mapValue
+	} else { //grouping
+		s.ClientResponses = append(s.ClientResponses, cr)
+	}
+}
+
+// HasNextClientResponses permits to verify if there are new client responses to be processed.
+func (s *SurveyStore) HasNextClientResponses() bool {
```

Comment:
Doesn't seem to be used - which is good. The comment and the name would propose the return-value to be `len(s.ClientResponses) > 0`.